### ISSUE TITLE
Determine if/what adjustments to the deployment strategy are needed.

### ISSUE THREAD

**Issue by adunkman** [ustaxcourt/ef-cms#254]
As the court, in order to have a successful deployment to the production environment with the latest code, we need to determine what our deployment strategy is.  

### Acceptance criteria:

- The Court has determined what is acceptable downtime. 
- Any changes needed are flushed out into Flexion’s backlog. 
- The deployment strategy is visualized with a diagram or other appropriate choice, keeping in mind that order of operations is important to illustrate.
- What is considered acceptable downtime and the rules around downtime are documented in the Site Operating Plan (#244).
- Users should not see errors related to a deployment being in progress
- Technical lead should be able to roll back to the previous deployment, including database (may be its own issue)
- Technical lead be able to automatically validate that the deployment was successful (may be its own issue)

### Notes: 

The Court has previously discussed as acceptance criteria: 

* Downtime should be minimized as much as possible.
* No downtime should be expected for routine updates.
    * Any downtime should be scheduled in advance, notifications are sent to users, and it should happen outside of 9-5pm business hours in all US timezones. 
    * Scheduled downtime should not exceed 24 hours.

After the main development phase, updates in production are expected 1-2 times per quarter.

* What kinds of changes cause downtime?
* Should version numbers been included in footers for customer support? 

Potentially related: 

- [Pink/Purple Deploys](https://trello.com/c/ewLCzN07/540-pink-purple-deploys)

**Comment by adunkman** [comment]
There’s still outstanding discussion here on what extent data migration scripts fit into the downtime requirements.

- How large of a data change will be a no-downtime deployment?
- When is the right time to develop code which handles larger data changes?

**Comment by adunkman** [comment]
Referencing [this comment](https://github.com/ustaxcourt/ef-cms/issues/241#issuecomment-651975441) from @ericsorenson (intended for this issue): 

> I've spoken to a few coworkers about approaches for reducing downtime, though applying some of these techniques will take engineering time and add AWS costs. We'll need to discuss options with the court to find a balance.

**Comment by adunkman** [comment]
Current unknowns: 

- How much downtime will routine deploys cause? 
  - #264, a test data set, would help to measure this.
- Where on the scale from "no downtime immediate deploys" to "24 hours of downtime" does the court want to sit, knowing that there is more effort associated with no-downtime immediate deploys?

**Comment by mmarcotte** [comment]
Bringing over from Flexion backlog: 
> The latest version of serverless-domain-manager is incompatible with our blue/green deploy system. We'll need a new approach to swap stacks after they've been smoke tested. 

Really this translates to nailing down the deployment strategy

**Comment by mmarcotte** [comment]
Have done some work on this [in this branch](https://github.com/ustaxcourt/ato/tree/deployments-documentation), but more work remains.

**Comment by mmarcotte** [comment]
Possible new issues for the backlog:
- [ ] User-friendly message for users while Dawson is down, with some mechanism to toggle this option and update the verbiage via a WYSIWYG editor. 
- [ ] Formalize Umzug migration script in the process taking into account realistic database size, blocked by #295 
- [ ] Build a mechanism to send out an email to all Court staff notifying them of downtime. Message should include a WYSIWYG editor for additional text to supplement a default stock message with the scheduled downtime. And an automatic reminder should be sent the day before the scheduled downtime.
- [ ] Rollback script for an Umzug migration: As a Court Tech Lead deploying new code that could change the database, I will need to have a testable rollback procedure for every deployment. The system should function and the database should match pre deployment. 

(work in progress)
- [ ]

**Comment by kfoley-18F** [comment]
Something to consider... past places I've worked have classified what kinds of changes require a downtime deploy and what kinds of changes don't require it. That allowed them to bundle all the downtime deploy changes together into a quarterly/less-frequent cycle, and still allow some changes to flow to production on a more frequent basis. Anything with a downtime deploy dependency is prioritized and managed in the backlog and branching structures differently to allow for faster iteration of non-downtime changes. It breaks the trunk branching model, but does offer a hybrid solution to still move along some work faster than an always-downtime-deploy strategy might allow.

**Comment by kfoley-18F** [comment]
@mmarcotte - ❤️  and 👍  on what's in your branch!
In the next iteration, it'd be helpful to have a "start" state on the deploy workflow diagrams - their circularity was confusing until I spent time with them.
